### DISCUSSION TITLE
Centralize environment variable handling

### DISCUSSION THREAD

**Discussion by zharinov** [renovatebot/renovate#24793] [Suggest an Idea]
**What would you like Renovate to be able to do?**

- Define the fixed set of variables we'll use and collect them (e.g. at env-config parsing phase)
- Allow for (re)definition from config (e.g. in `packageRules`)

**Did you already have any implementation ideas?**

Maybe, some kind of file-scoped cache encapsulated by exported `getEnvVar` and `setEnvVar` functions.

**Comment by viceice** [comment]
I would prefer to combine this with our `customEnvVariables` feature. 

 https://docs.renovatebot.com/self-hosted-configuration/#customenvvariables

**Comment by rarkins** [comment]
I think long term we want to allow users to supply env variables in an Object similar to `aliases` for helm charts. Ideally they are as configurable as possible like most of our config options, down to the per-package level.

However we'll also need to allow bot admins to whitelist variable names and patterns, because some env variables could be used for malicious purposes.

Any propose solution should keep in mind this future user-provided env variable goal.

**Comment by rarkins** [comment]
I'm thinking about the following approach:

- A new user config option called e.g. `env`, which is `Record<string, string>`
- A new admin config called e.g. `allowedEnv` which is an array of regex patterns to "whitelist" allowed env. If an env value is allowed and specified by the user then it would override anything from the real env
- This `env` value is passed to manager during extract and artifacts update
- Each manager might know env which is save to use/pass to package managers, e.g. `GONOPROXY`, but anything matching `allowedEnv` and present in `env` would be passed through to package managers too

Decision:
- Should `env` be top-level only or configurable per-package using `packageRules`? If the latter then it means we need to do the whitelisting "late" and also work out how to handle when values conflict within multiple upgrades in the same config. environment variables are normally fixed, not changing per-package..

Questions:
- Do we have a use-case for this?
- How does this combine with `customEnvVariables`?
